### PR TITLE
[Environment Variable][3/N] Use thread-safe getenv wrapper

### PR DESCRIPTION
**Pull request by cyyever**
Follows #124485

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by colesbury** [comment]
These things are only thread-safe if *all* the calls are under the same lock and that's not likely to be the case. For example, uses of the `getenv` from Python or in our third-party libraries won't use this.

**Comment by cyyever** [comment]
Of course, these functions should be applied consistently across the whole code base to achieve the effects.

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -3130,9 +3131,10 @@ class DeviceCachingAllocator {
 static bool forceUncachedAllocator() {
   // Allow either CUDA or HIP name for env var for maximum user comfort
   // the CUDA env var avoids being hipified in cuda_to_hip_mappings.py
-  static const char* cuda_env = getenv("PYTORCH_NO_CUDA_MEMORY_CACHING");
-  static const char* rocm_env = getenv("PYTORCH_NO_HIP_MEMORY_CACHING");
-  static bool force_uncached = (cuda_env != nullptr) || (rocm_env != nullptr);
+  static bool cuda_env = c10::utils::has_env("PYTORCH_NO_CUDA_MEMORY_CACHING");
+  static bool rocm_env = c10::utils::has_env("PYTORCH_NO_HIP_MEMORY_CACHING");
+  static bool force_uncached = cuda_env || rocm_env;
+ main
```
Comment (eqy) [inline c10/cuda/CUDACachingAllocator.cpp:3137]: Is this a typo?

**Comment by cyyever** [comment]
@eqy Fixed incorrect rebase.

File: c10/cuda/CUDACachingAllocator.cpp (hunk 1 above)
Comment (cyyever) [reply c10/cuda/CUDACachingAllocator.cpp]: Typo, fixed,

**Review by eqy — APPROVED** [review]

**Pull request closed without merging**